.github/workflows/integration.yml

```diff
@@ -58,17 +64,111 @@ jobs:
-      - name: Start sg-bridge for collectd from container image
+      - name: Start sg-bridge from container image
         if: steps.bridge_branch.outcome != 'success'
         run: |
           docker run --name=sgbridge --network host $BRIDGE_VOLUME -d \
-            $BRIDGE_IMAGE --amqp_url amqp://localhost:5666/$QDR_CHANNEL_COLLECTD \
+            $BRIDGE_IMAGE --amqp_url amqp://localhost:5666/$QDR_CHANNEL \
             --gw_unix=$BRIDGE_SOCKET
-      - name: Start sg-bridge for collectd with same branch
+      - name: Start sg-bridge for with same branch
         if: steps.bridge_branch.outcome == 'success'
         run: |
           docker run --name=sgbridge --network host $BRIDGE_VOLUME -d -uroot \
-            -e GITHUB_REF -e BRIDGE_SOCKET -e QDR_CHANNEL_COLLECTD -e OPSTOOLS_REPO \
+            -e GITHUB_REF -e BRIDGE_SOCKET -e QDR_CHANNEL -e OPSTOOLS_REPO \
+            --workdir=$(dirname $BRIDGE_SOCKET) \
+            $TEST_IMAGE bash $PROJECT_ROOT/ci/integration/metrics/run_bridge.sh
+      - name: Install collectd
+        run: |
+          sudo apt-get install collectd
+          sudo systemctl stop collectd && sudo systemctl disable collectd
+          sudo cp ci/integration/metrics/collectd/collectd.conf /etc/collectd/collectd.conf
+          sudo touch /var/log/collectd.log && sudo chmod a+rw /var/log/collectd.log
+          sudo collectd -C ci/integration/metrics/collectd/collectd.conf
+      - name: Run sg-core to process metrics
+        run: |
+          docker run --name=sgcore -d -uroot --network host $BRIDGE_VOLUME -e OPSTOOLS_REPO \
+            --volume ${{ github.workspace }}:$PROJECT_ROOT:z --workdir $PROJECT_ROOT \
+            $TEST_IMAGE bash $PROJECT_ROOT/ci/integration/metrics/collectd/run_sg.sh
+      - name: Run Prometheus to store metrics
+        run: |
+          docker run --name=prometheus -d --network host \
+          --volume ${{ github.workspace }}/ci/integration/metrics/prometheus.yml:/etc/prometheus/prometheus.yml:ro \
+          $PROMETHEUS_IMAGE
+      - name: Debug output
+        run: |
+          sleep 360
+          echo "=========================== qdr =========================" && \
+          docker exec qdr qdstat -b 127.0.0.1:5666 -a
+          docker logs qdr
+          echo "========================= sg-core =======================" && \
+          docker logs sgcore
+          echo "======================== prometheus =====================" && \
+          docker logs prometheus
+      - name: Validate metrics processing
+        run: |
+          docker run --name=validate -uroot --network host \
+            --volume ${{ github.workspace }}:$PROJECT_ROOT:z --workdir $PROJECT_ROOT \
+            $TEST_IMAGE bash $PROJECT_ROOT/ci/integration/metrics/collectd/run_validation.sh
+#-------------------------------------------------------------------------------
+  ceilometer-metrics-bridge:
```

Comment:
Just noting that further refactoring could happen here using composite actions (https://docs.github.com/en/actions/creating-actions/creating-a-composite-action) in order to remove some of this duplication.

I am not advocating doing it here though, just noting in case a future refactoring was warranted.

Reply:
Yes I agree. For the metric jobs for example, the first ~5 steps are the same. I was looking into the composite actions but I decided to not implement them. As I understand the documentation, we would either need to create a new repository for the actions or we could have them in this repo, but we would probably need to tag the repository. So that we can use the `uses: action@tag` syntax. I haven't tried this, but after reading the docs and understanding it like this, I decided to not use the composite actions to not create some confusion in the repository tags just for CI deduplication.